pocs/camera/__init__.py

```diff
@@ -115,71 +115,73 @@ def kwargs_or_config(item, default=None):
 
     primary_camera = None
 
+    # Different models require different connections methods.
+    model_requires = {
+        'canon_gphoto2': 'port',
+        'sbig': 'serial_number',
+        'zwo': 'serial_number',
+        'fli': 'serial_number',
+    }
+
     device_info = camera_info['devices']
     for cam_num, device_config in enumerate(device_info):
         cam_name = 'Cam{:02d}'.format(cam_num)
 
         if not a_simulator:
-            camera_model = device_config.get('model')
-
             # Assign an auto-detected port. If none are left, skip
             if auto_detect:
                 try:
-                    camera_port = ports.pop()
+                    device_config['port'] = ports.pop()
                 except IndexError:
-                    logger.warning(
-                        "No ports left for {}, skipping.".format(cam_name))
+                    logger.warning("No ports left for {}, skipping.".format(cam_name))
                     continue
             else:
+                # Check for proper connection method.
+                model = device_config['model']
                 try:
-                    camera_port = device_config['port']
-                except KeyError:
+                    connection_method = model_requires[model]
+                    if connection_method not in device_config:
+                        raise error.CameraNotFound
```

Comment:
Is this the right error here? If I'm following correctly this would be the result of a bad configuration (device config missing the port/serial number it needs), rather than an expected camera not being present.

Reply:
More to the point, I'm not entirely sure we should be raising an exception inside a for loop. This means, in the case of Huntsman, if one of the cameras has problems then all 10 would fail to initialise.

What we are actually testing here is the `create_cameras_from_config` method, which will eventually need to be overhauled for distributed cameras and for a change in the config server, rather than the actual camera creation process itself.

I'll probably end up changing this test a bit.